flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests the metrics for input buffers usage.
+ */
+public class InputBuffersMetricsTest extends TestLogger {
+
+	@Test
+	public void testBufferUsageMetrics() throws IOException, InterruptedException {
+
+		int numberOfRemoteChannelsGate1 = 2;
+		int numberOfLocalChannelsGate1 = 0;
+		int numberOfRemoteChannelsGate2 = 1;
+		int numberOfLocalChannelsGate2 = 1;
+
+		int numberOfRemoteChannelsTotal = numberOfRemoteChannelsGate1 + numberOfRemoteChannelsGate2;
+		int numberOfInputGates = 2;
+
+		int buffersPerChannel = 2;
+		int extraNetworkBuffersPerGate = 8;
+
+		final CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+		final NettyShuffleEnvironment network = new NettyShuffleEnvironmentBuilder()
+			.setNetworkBuffersPerChannel(buffersPerChannel)
+			.setFloatingNetworkBuffersPerGate(extraNetworkBuffersPerGate)
+			.build();
+
+		final Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(network,
+			numberOfRemoteChannelsGate1,
+			numberOfLocalChannelsGate1);
+		final Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple2 = buildInputGate(network,
+			numberOfRemoteChannelsGate2,
+			numberOfLocalChannelsGate2);
+
+		final SingleInputGate inputGate1 = tuple1.f0;
+		final SingleInputGate inputGate2 = tuple2.f0;
+
+		try {
+
+			assertEquals(tuple1.f1.size(), numberOfRemoteChannelsGate1 + numberOfLocalChannelsGate1);
+			RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
+			RemoteInputChannel remoteInputChannel2 = tuple1.f1.get(1);
+
+			SingleInputGate[] inputGates = new SingleInputGate[]{tuple1.f0, tuple2.f0};
+			FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBuffersUsageGauge inputBufferPoolUsageGauge = new CreditBasedInputBuffersUsageGauge(floatingBuffersUsageGauge,
+				exclusiveBuffersUsageGauge, inputGates);
+
+			assertEquals(0.0, floatingBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, exclusiveBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBufferPoolUsageGauge.getValue(), 0.0);
+
+			assertEquals(buffersPerChannel * numberOfRemoteChannelsGate1, exclusiveBuffersUsageGauge.calculateTotalBuffers(inputGate1));
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateTotalBuffers(inputGate1));
+
+			assertEquals(buffersPerChannel * numberOfRemoteChannelsGate2, exclusiveBuffersUsageGauge.calculateTotalBuffers(inputGate2));
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateTotalBuffers(inputGate2));
```

Comment:
If we write a single test only for verifying the function of `FloatingBuffersUsageGauge#calculateTotalBuffers`, then we only need one `SingleInputGate`, do not need to verify it twice for both `inputGate1` and `inputGate2`.

The array of `SingleInputGate` is only reasonable for verifying the function of `AbstractBuffersUsageGauge#getValue`.

Reply:
Yes, I will break this test case into serval independent methods.